File: Core/Export/ExportService.swift
import AppKit
import Foundation
import UniformTypeIdentifiers

enum ExportService {
    @MainActor
    static func saveCSV(devices: [Device]) throws {
        let data = csvData(for: devices)
        try save(data: data, suggestedName: "lanscope-devices.csv", contentType: .commaSeparatedText)
    }

    @MainActor
    static func saveJSON(devices: [Device]) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(devices)
        try save(data: data, suggestedName: "lanscope-devices.json", contentType: .json)
    }

    @MainActor
    static func copyTSV(devices: [Device]) {
        let headers = ["Status", "Name", "IP Address", "MAC Address", "Vendor", "Open Ports", "Services", "Last Seen"]
        let rows = devices.map { device in
            [
                device.status.title,
                device.displayName,
                device.ipAddress,
                device.macAddress ?? "",
                device.vendor,
                device.openPortsDisplay,
                device.servicesDisplay,
                DateFormatter.lanScopeDateTime.string(from: device.lastSeen)
            ].joined(separator: "\t")
        }
        let value = ([headers.joined(separator: "\t")] + rows).joined(separator: "\n")
        DeviceActionService.copy(value)
    }

    static func csvString(for devices: [Device]) -> String {
        let headers = ["Status", "Name", "IP Address", "MAC Address", "Vendor", "Open Ports", "Services", "Last Seen"]
        let rows = devices.map { device in
            csvRow([
                device.status.title,
                device.displayName,
                device.ipAddress,
                device.macAddress ?? "",
                device.vendor,
                device.openPortsDisplay,
                device.servicesDisplay,
                DateFormatter.lanScopeDateTime.string(from: device.lastSeen)
            ])
        }
        return ([csvRow(headers)] + rows).joined(separator: "\n") + "\n"
    }

    static func csvData(for devices: [Device]) -> Data {
        Data(csvString(for: devices).utf8)
    }

    private static func csvRow(_ values: [String]) -> String {
        values.map { value in
            "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
        }
        .joined(separator: ",")
    }

    @MainActor
    private static func save(data: Data, suggestedName: String, contentType: UTType) throws {
        let panel = NSSavePanel()
        panel.nameFieldStringValue = suggestedName
        panel.allowedContentTypes = [contentType]
        panel.canCreateDirectories = true

        guard panel.runModal() == .OK, let url = panel.url else {
            return
        }

        try data.write(to: url)
    }
}

File: Windows/LanScope.Core/Services/ExportService.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using LanScope.Core.Models;

namespace LanScope.Core.Services;

public static class ExportService
{
    public static string ToCsv(IEnumerable<Device> devices)
    {
        var rows = new List<string>
        {
            CsvRow(["Status", "Name", "IP Address", "MAC Address", "Vendor", "Open Ports", "Services", "Last Seen"])
        };
        rows.AddRange(devices.Select(device => CsvRow([
            device.StatusDisplay,
            device.DisplayName,
            device.IpAddress,
            device.MacAddress ?? "",
            device.Vendor,
            device.OpenPortsDisplay,
            device.ServicesDisplay,
            device.LastSeen.ToString("O")
        ])));
        return string.Join(Environment.NewLine, rows) + Environment.NewLine;
    }

    public static string ToJson(IEnumerable<Device> devices) => JsonSerializer.Serialize(devices, new JsonSerializerOptions
    {
        WriteIndented = true,
        Converters = { new JsonStringEnumConverter() }
    });

    public static string ToTsv(IEnumerable<Device> devices)
    {
        var rows = new List<string> { "Status\tName\tIP Address\tMAC Address\tVendor\tOpen Ports\tServices\tLast Seen" };
        rows.AddRange(devices.Select(x => string.Join('\t', [
            x.StatusDisplay, x.DisplayName, x.IpAddress, x.MacAddress ?? "", x.Vendor,
            x.OpenPortsDisplay, x.ServicesDisplay, x.LastSeen.ToString("g")
        ])));
        return string.Join(Environment.NewLine, rows);
    }

    public static void SaveUtf8(string path, string content) => File.WriteAllText(path, content, new UTF8Encoding(true));
    private static string CsvRow(IEnumerable<string> values) => string.Join(',', values.Select(x => $"\"{x.Replace("\"", "\"\"")}\""));
}
